internal/appconfig/watcher.go

```diff
@@ -23,25 +24,245 @@ import (
+// You must call Close to properly clean up resources used by the Watcher.
+func NewWatcher(opts ...Option) (*Watcher, error) {
+	var bgWg sync.WaitGroup
+	bgCtx, bgCancel := context.WithCancel(context.Background())
+
+	// If we return due to an error, cancel the background context.
+	// This won't do anything on success cause we nil out bgCancel.
+	defer func() {
+		if bgCancel != nil {
+			bgCancel()
+		}
+	}()
+
+	// Build our initial watcher
+	w := &Watcher{
+		log:             hclog.L(),
+		dynamicEnabled:  true,
+		refreshInterval: defaultRefreshInterval,
+		plugins:         map[string]*plugin.Instance{},
+		inSourceCh:      make(chan []*pb.ConfigSource),
+		inVarCh:         make(chan []*pb.ConfigVar),
+		currentCond:     sync.NewCond(&sync.Mutex{}),
+		bgCtx:           bgCtx,
+		bgCancel:        bgCancel,
+		bgWg:            &bgWg,
+	}
+
+	// Use the option pattern to update any options.
+	for _, opt := range opts {
+		if err := opt(w); err != nil {
+			return nil, err
+		}
+	}
+
+	// Start our background goroutine
+	w.bgWg.Add(1)
+	go w.watcher(
+		bgCtx,
+		w.log.Named("watchloop"),
+	)
+
+	// Everything is good, nil out bgCancel so our defer doesn't stop us
+	bgCancel = nil
+
+	return w, nil
+}
+
+// Close stops all the background goroutines that this watcher started.
+// This will block until all the background tasks have exited.
+func (w *Watcher) Close() error {
+	w.bgCancel()
+	w.bgWg.Wait()
+	return nil
+}
+
+// Next returns the next values for the configuration AFTER the given
+// iterator value iter. A value of 0 can be used for iter for a first read.
+//
+// The return value will be the configuration values in env format (KEY=VALUE),
+// the current iterator value that you should use with the next call to Next,
+// and any error if it occurred.
+//
+// The ctx parameter can be used for timeouts, cancellation, etc. If the context
+// is closed, this will return the context error.
+func (w *Watcher) Next(ctx context.Context, iter uint64) ([]string, uint64, error) {
+	var doneCh chan struct{}
+
+	w.currentCond.L.Lock()
+	defer w.currentCond.L.Unlock()
+
+	// Wait on the condition var as long as we have the same iterator
+	// and the context isn't yet cancelled.
+	for w.currentGen == iter && ctx.Err() == nil {
+		// If we're waiting, then we want to start a goroutine to notify
+		// us if the context closes. We have to do this in a goroutine because
+		// cond vars have no other way to wait on a context.
+		//
+		// We do this in the for loop so that on the fast path where we
+		// have an older generation, we just return the value immediately
+		// without all the goroutine ceremony.
+		if doneCh == nil {
+			// doneCh ensures that we clean up our goroutines when we return.
+			doneCh = make(chan struct{})
+			defer close(doneCh)
+
+			go func() {
+				select {
+				case <-ctx.Done():
+					// Wake up all condition vars so we wake ourself up.
+					w.currentCond.Broadcast()
+
+				case <-doneCh:
+					// Return
+				}
+			}()
+		}
```

Comment:
More a future observation that pull this functionality out to allow a condvar to integrate with channels would be good generic functionality.